### PR TITLE
Implement `choice_{to, from}_index`

### PR DESCRIPTION
**Pull request by tybug**
Depends on #4208. 

This is a core chunk-of-work which marks the foundation of starting to finish off the typed choice sequence migration. It introduces two primitives: `choice_to_index` and `choice_from_index`. These can be used to build up to and replace buffer-based definitions, like shrink ordering (order by index) and generating zero-buffer aka simplest inputs (`choice_from_index(0)`).

These functions are only currently used in datatree for generating hard values, and in computing trivial/simplest values. Shrinking is still on the buffer, for now.

I've added a new file `choice.py`, which I envision as the new home for what we previously called the "ir" and which currently lives in `data.py`: `IRNode` -> `ChoiceNode`? just `Choice`?, `NodeTemplate` -> `ChoiceTemplate`, `ir_size` -> `choice_size`, etc. Since this file is confusable with the totally separate `choicetree.py`, I've moved that into the shrinking subfolder.

### CODE REVIEW

**Comment by tybug** [comment]
(merged to make the diff more palatable, apologies for any churn)

File: hypothesis-python/src/hypothesis/internal/conjecture/choice.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        #   {min, max}_value. so min_value=-1 a=0 has order
+        #   [0, 1, -1, 2, 3, 4, ...]
+        # * Bounded: Ordered by (sgn(a - x), |a - x|). Count upwards until max_value,
+        #   then count downards.
+        #   [a, a + 1, a + 2, ..., max_value, a - 1, a - 2, ..., min_value]
+        #
+        # To simplify and gain intuition about this ordering, you can think about
+        # the most common case where 0 is first (a = 0). We deviate from this only
+        # rarely, e.g. for datetimes, where we generally want year 2000 to be
+        # simpler than year 0.
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/conjecture/choice.py:154]: I think we want to use the same comparison order for bounded intervals as for unbounded or semi-bounded.

**Review by Zac-HD — APPROVED** [review]
I'd like to use a consistent shrinking order for all integers choices, as below, but overall this looks great - and I'm excited that we're nearing the final stages!

File: hypothesis-python/src/hypothesis/internal/conjecture/choice.py (hunk 1 above)
Comment (tybug) [reply hypothesis-python/src/hypothesis/internal/conjecture/choice.py]: Thanks for catching this, have done so. I think I was trying to match the existing ordering at the time (for no principled reason).

**Pull request merged**